src/charm.py

```diff
@@ -9,93 +9,306 @@
+
+        if removed_relations:
+            # update configuration
+            self._on_config_changed(event)
+            # update relation status to removed
+            for relation in removed_relations:
+                self.set_secret("app", relation, "removed")
+
+    @property
+    def topic_name(self) -> Optional[str]:
+        """Return the configured topic name."""
+        return self.model.config.get("topic-name", None)
+
+    @property
+    def database_name(self) -> Optional[str]:
+        """Return the configured database name."""
+        return self.model.config.get("database-name", None)
+
+    @property
+    def extra_user_roles(self) -> Optional[str]:
+        """Return the configured extra user roles."""
+        return self.model.config.get("extra-user-roles", None)
+
+    @property
+    def kafka_relation(self) -> Optional[Relation]:
+        """Return the kafka relation if present."""
+        return self.kafka.relations[0] if len(self.kafka.relations) > 0 else None
+
+    @property
+    def database_relations(self) -> Dict[str, Relation]:
+        """Return the active database relations."""
+        return {
+            name: requirer.relations[0]
+            for name, requirer in self.databases.items()
+            if len(requirer.relations) > 0
+        }
+
+    @property
+    def databases_active(self) -> Dict[str, str]:
+        """Return the configured database name."""
+        return {
+            name: relation.data[self.app]["database"]
+            for name, relation in self.database_relations.items()
+            if "database" in relation.data[self.app]
+        }
+
+    @property
+    def topic_active(self) -> Optional[str]:
+        """Return the configured topic name."""
+        if relation := self.kafka_relation:
+            if "topic" in relation.data[self.app]:
+                return relation.data[self.app]["topic"]
+        return None
+
+    @property
+    def extra_user_roles_active(self) -> Optional[str]:
+        """Return the configured user-extra-roles parameter."""
+        return (
+            relation.data[self.app]["extra-user-roles"]
+            if (relation := self.kafka_relation)
+            else None
+        )
+
+    @property
+    def is_database_related(self) -> bool:
+        """Return if a relation with database is present."""
+        possible_relations = [
+            self._check_for_credentials(database_requirer.relations)
+            for _, database_requirer in self.databases.items()
+        ]
+        return any(possible_relations)
+
+    @staticmethod
+    def _check_for_credentials(relations: List[Relation]) -> bool:
+        """Check if credentials are present in the relation databag."""
+        for relation in relations:
+            if (
+                "username" in relation.data[relation.app]
+                and "password" in relation.data[relation.app]
+            ):
+                return True
+        return False
```

Comment:
**question:** I'm sure it's fine, but just checking, do we mean to exit on first relation found with username/password? 

What about the other relations? Or is it assumed that any list we pass this function, if one is set, they all are?

Reply:
I believe we have only 1 relation per type. See the metadata.yaml

https://github.com/canonical/data-integrator/blob/main/metadata.yaml

But I'm not sure what Juju does if we try to relate multiple times on the same interface. Would it complain?

Reply:
Yes It will complain, Juju will allow only one relation per interface.